content/blog/2020-12-29-gator-oop/index.md

```diff
@@ -0,0 +1,315 @@
+
+## Main Design Decisions
+
+Since GLSL has structs and functions, I thought a reasonable way to implement classes would be to strip out all the fields and place them into structs, and implement methods as functions which consume these "field structs".
+
+I decided to implement single inheritance. In addition, all fields and methods have one of three access modifiers: public, private, or protected.
+
+Since the target language, GLSL, does not contain function pointers, Gator cannot support dynamic dispatch. Therefore, valid Gator classes cannot define a method which is already defined by a parent class. At the moment, this is actually not enforced by the compiler.
+
+## Constructor Fun
+
+I originally designed constructors to work as follows. It should be noted that some of the syntax can be improved, particularly for constructors. CS 6120 is not about parsing ;)
+
+```
+class dog : animal {
+    private float bark_volume;
+
+    public dog init(float bv, int x, int y) {
+        super(x, y);
+        this.bark_volume = bv;
+    }
+};
+```
+
+My initial idea was to implement this by creating an underlying GLSL dog struct with default initial values. Then, I would set the `bark_volume` field to `bv` and the implicit `super` field to the result of calling the constructor for `animal`. This did not work because there was not a reasonable way to populate the fields with default values. For integers or floats, I could of course use zero, but for fields which are other objects for example, this approach becomes difficult.
+
+The redesigned constructors work as follows. If the class has a parent, the first line must still be a call to `super`, but now the last line of the constructor must be a call to `self`, with the fields provided in the order they are declared in the class.
+
+```
+class dog : animal {
+    private float bark_volume;
+
+    public dog init(float bv, int x, int y) {
+        super(x, y);
+        self(bv);
+    }
+};
+```
+
+This way the underlying struct can be initialized at the end, when all field values are known, including the implicit super struct.
+
+The above example indirectly calls the constructor for `animal`, but constructors can also be directly called as follows.
+
+```
+dog quiet_dog_factory(int x, int y) {
+   return new_dog(0.25, x, y);
+}
+```
+
+Idiotically, I realized when I performed the evaluation that there was a way to use my first constructor approach. Apparently, although GLSL doesn't provide default constructors for structs, you can simply declare a struct and then set the fields, without ever using a GLSL structure constructor.
+
+## Sample Output
+
+Below are some test Gator programs and their corresponding GLSL output.
```

Comment:
Perhaps there is a way to distill these examples down to their essence, so it's possible for readers to look at exactly what you want them to look at? A couple of sentences about the takeaways people are supposed to get out of these examples would be helpful too. Or, if there's not really much to say and you're just indicating that it works, maybe it would be better to just link to the example files on GitHub so the super-curious can read them, without interrupting the flow of exposition for normal readers.